[PIP-129] Introduce intermediate state for ledger deletion

wuzhanpeng opened the pull request:
Master Issue: #13526 

### Motivation

This pull request is the specific implementation plan of PIP-129

### Modifications

- Mark the deletable ledgers before actually delete them
- Add the ability to check and track deletable ledgers

Most changes are concentrated on `org.apache.bookkeeper.mledger.ManagedLedger` and `org.apache.bookkeeper.mledger.impl.ManagedLedgerImpl`.

### Verifying this change

- [x] Make sure that the change passes the CI checks.

This change is already covered by existing tests, such as *managed-ledger*.

This PR is just a refactoring of the ledger deletion. The original logic semantics has not changed. I think the existing test cases of unit test and integration test are sufficient for test coverage.

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation
  
- [x] `no-need-doc` 
  
Only involves the improvement and reconstruction of internal logic.

mattisonchao commented:
@wuzhanpeng 

Hi, wuzhanpeng

Can you help me understand the meaning of WIP in Pulsar?

wuzhanpeng commented:
/pulsarbot run-failure-checks

wuzhanpeng commented:
/pulsarbot run-failure-checks

wuzhanpeng commented:
/pulsarbot run-failure-checks

wuzhanpeng commented:
@codelipenghui @hangc0276 Could you help review this PR?

wuzhanpeng commented:
/pulsarbot run-failure-checks

wuzhanpeng commented:
/pulsarbot run-failure-checks

wuzhanpeng commented:
/pulsarbot run-failure-checks

wuzhanpeng commented:
/pulsarbot run-failure-checks

codelipenghui commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java:
```diff
@@ -649,4 +651,31 @@ void asyncSetProperties(Map<String, String> properties, AsyncCallbacks.UpdatePro
      * @return the future of managed ledger internal stats
      */
     CompletableFuture<ManagedLedgerInternalStats> getManagedLedgerInternalStats(boolean includeLedgerMetadata);
+
+    /**
+     * Mark deletable ledgers for bookkeeper and offload storage.
+     *
+     * @param deletableLedgerIds
+     * @param deletableOffloadedLedgerIds
+     */
+    void markDeletableLedgers(Collection<Long> deletableLedgerIds, Collection<Long> deletableOffloadedLedgerIds);
+
+    /**
+     * Get all deletable ledgers.
+     *
+     * @return all the deletable ledgers of the managed-ledger
+     */
+    Set<Long> getAllDeletableLedgers();
+
+    /**
+     * Get all deletable offloaded ledgers.
+     *
+     * @return all the deletable offloaded ledgers of the managed-ledger
+     */
+    Set<Long> getAllDeletableOffloadedLedgers();
+
+    /**
+     * Check and remove all the deletable ledgers.
+     */
+    void removeAllDeletableLedgers();
```
Do we need to expose these methods to the ManagedLedger interface? Looks they are only used in the internal of the ManagedLedgerImpl.

hangc0276 replied:
+1

hangc0276 replied:
+1

hangc0276 commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:
```diff
@@ -2513,7 +2531,25 @@ void internalTrimLedgers(boolean isTruncate, CompletableFuture<?> promise) {
+
+            // Mark deletable offloaded ledgers
+            Set<Long> deletableOffloadedLedgers = ledgersToDelete.stream()
+                    .filter(ls -> ls.getOffloadContext().hasUidMsb())
```
You'd better check `ls.getOffloadContext().isComplete` instead of `ls.getOffloadContext().hasUidMsb()`

wuzhanpeng replied:
This is mainly implemented according to the original logic. See https://github.com/apache/pulsar/blob/c18d64526708c6dd051217feba3c31a73507620c/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java#L2716-L2729

hangc0276 commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:
```diff
@@ -4039,6 +4122,190 @@ public void clearBacklogFailed(ManagedLedgerException exception, Object ctx) {
+
+            if (allFailedLedgers.isEmpty()) {
+                log.info("[{}] ledgers: {} and offloaded ledgers: {} are deleted successfully.",
+                        name, deletableLedgers, deletableOffloadedLedgers);
```
`deletableLedgers` and `deletableOffloadedLedgers` are Set and not override `toString`, So it will print the object address instead of the items.

wuzhanpeng replied:
Those collection objects that inherit `java.util.AbstractCollection` have the ability to return a string representation of this collection (`java.util.AbstractCollection#toString`).

hangc0276 commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:
```diff
@@ -4039,6 +4122,190 @@ public void clearBacklogFailed(ManagedLedgerException exception, Object ctx) {
+            if (log.isDebugEnabled()) {
+                log.debug("[{}] Successfully delete bookkeeper ledgers: {} and offloaded ledgers: {}. "
+                                + "Failed to delete bookkeeper ledgers: {} and offloaded ledgers: {}", name,
+                        succeedDeletedLedgers, succeedDeletedOffloadedLedgers,
```
The same above.

hangc0276 commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:
```diff
@@ -2547,21 +2582,15 @@ void internalTrimLedgers(boolean isTruncate, CompletableFuture<?> promise) {
                 @Override
                 public void operationComplete(Void result, Stat stat) {
+                    // perform actual deletion
+                    removeAllDeletableLedgers();
```
it will block at most 30s to wait all the ledger delete operation completed. However, this method is called by meta store callback thread, which will block the thread and lead to deadlock.

wuzhanpeng replied:
The reason for performming `removeAllDeletableLedgers` synchronously here is that two locks of `trimmer` and `metadata` need to be released after the actual deletion. If the execution is asynchronous, the update process of the property map cannot be protected.

hangc0276 replied:
The operationComplete will be called by `bookkeeper-ml-scheduler` thread pool, If it executed by synchronously, it will block other operations. We can use another new thread pool the execute it. 
For the property map problem, could we use another thread safe data structure the store it?

hangc0276 commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:
```diff
@@ -146,6 +151,13 @@
 
     protected static final int AsyncOperationTimeoutSeconds = 30;
 
+    protected static final String DELETABLE_LEDGER_MARKER_KEYWORD = "pulsar.ml.deletable.ledgers";
```
We will append the prefix for each ledger, and store the property map into ZNode. However, if there are thousands of ledgers to delete, the property map will cost too many storage space especially into Znode, which limit by 5MB. The key point is the prefix will duplicate for each ledger. We'd better redesign the data structure of the properties storage.

wuzhanpeng replied:
Thanks for your question. IMO, the problem of over-size znode should theoretically be a discussion about whether `ManagedLedgerInfo` should be stored on zk. Or we can think of it this way, the ledger information stored in the property map actually comes from the `LedgerInfoList` structure, and this part of the data is also stored in a znode. Therefore, if the number of ledgers is large, the corresponding ml-znode must be huge. To avoid complicating matters, I think we can skip this part in this proposal, and start a new discussion thread on this problem.

hangc0276 replied:
For current implementation, we shouldn't introduce more overload for current managedLedger Znode storage. we'd better reduce the prefix of keyword or redesign the data structure for properties storage.

zymap replied:
Maybe we keep the `d_ledgers` is enough? The `pulsar.ml` looks useless.

hangc0276 reviewed (changes requested):
Please add tests for this feature to protect the ledger deletion logic. @wuzhanpeng

hangc0276 commented:
@wuzhanpeng We'd better expose the ledger delete state into metrics.

wuzhanpeng commented:
According to the review comments, I have made the following modifications: 
- removed the related ledger deletion methods from the managed-ledger interface
- added related unit tests for two-stage ledger deletion 
- added monitoring metrics during ledger deletion

@codelipenghui @hangc0276 PTAL at your convenience~

wuzhanpeng commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:
```diff
@@ -2718,42 +2747,52 @@ public void deleteCursorFailed(ManagedLedgerException exception, Object ctx) {
+        asyncDeleteLedger(ledgerId, retry, new DeleteLedgerCallback() {
+            @Override
+            public void deleteLedgerComplete(Object ctx) {
+
```
This is the default implementation, which is used to be compatible with the original logic.

wuzhanpeng commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:
```diff
@@ -2523,7 +2541,25 @@ void internalTrimLedgers(boolean isTruncate, CompletableFuture<?> promise) {
             // Update metadata
+            // Mark deletable ledgers
+            Set<Long> deletableLedgers = Stream
+                    .concat(
```
There may be ledgers in both `ledgersToDelete` and `offloadedLedgersToDelete` that need to be deleted. They do have the potential to have duplicate ledgers and that's why `toSet` is used.

wuzhanpeng commented:
I have made some updates based on the comments, PTAL~ @hangc0276 @zymap

wuzhanpeng commented:
/pulsarbot run-failure-checks

wolfstudy commented:
What is the stats of the change? @wuzhanpeng

tisonkun commented:
Closed as stale and conflict. Please rebase and resubmit the patch if it's still relevant.
